### ISSUE TITLE
Compile error with cloneAndMutateFunc use for creation of a Store

### ISSUE THREAD

**Issue by st-clair-clarke** [giusepperaso/structura.js#108]
I have the following Typescript types and classes:

```

import { nanoid } from 'nanoid'

import { immerable } from 'immer'

export type TID = string | number

export type TAge = {
   id: TID

   dob: Date
   dobTime: Date

   dod: Date
   dodTime: Date

   years: number
   months: number
   weeks: number
   days: number
   hours: number
   minutes: number
}

export class Age {

   id: TID = nanoid()

   dob: Date = new Date()

   dobTime: Date = new Date()

   dod: Date = new Date()

   dodTime: Date = new Date()

   years = 0

   months = 0

   weeks = 0

   days = 0

   hours = 0

   minutes = 0
}


import { produce } from "structurajs"

@Injectable({ providedIn: 'root' })
export class AgeStore extends ImmutableStore<TAge> {
  constructor() {
    super( {
             cloneAndMutateFunc: produce, // COMPILE ERROR
             enableLogging: true,
             initialState: new Age(),
             name: 'RegistrationStore',
             plugins: [
               useDevtools(),
               useStoreHistory(),
               useStorePersistence(),
             ],
           } )
  }
}


```

Compile error:

```
<T, Q, IS_ASYNC = false>(state: DraftableState<T>, producer: Producer<T, Q, IS_ASYNC>, patchCallback?: PatchCallback<T> | undefined, { proxify }?: ProduceOptions | undefined) => ProduceReturn<...>
is not assignable to type
( currentState: TAge, mutation: ( draftState: TAge ) => void ) => TAge
Types of parameters  producer  and  mutation  are incompatible.
Types of parameters  draftState  and  draft  are incompatible.
Type  UnFreezedObject<TAge>  is not assignable to type  TAge 
The types of  dob.toString  are incompatible between these types.
Type  UnFreezedObject<() => string>  is not assignable to type  () => string 
Type  UnFreezedObject<() => string>  provides no match for the signature  (): string 
immutable-store-config.d.ts(16, 5): The expected type comes from property  cloneAndMutateFunc  which is declared here on type  ImmutableStoreConfig<TAge> 
```

If I use immer, the error disappears. However, I would like to use structurajs instead. Appreciate your help in correcting the error.

**structrajs DOES NOT KNOW HOW TO PARSE Date type**

Cheers

**Comment by zuriscript** [comment]
This is solely a type inference issue, and can be solved by the client of both libraries.
Structura.js `produce` type definition can not be directly matched to the geneirc base type `(currentState: TState, mutation: (draftState: TState) => void) => TState`. The inference works with immer.js, but structura.js offers more features with it's type system, which might complicate things.

Solution:
You could either use a proxy function to wire the arguments with the right types or simply use double assertions (since structura.js actually supports the same basis signature as immer.js has):

```typescript
export type MutationFn<T> = (currentState: T, mutation: (draftState: T) => void) => T;

@Injectable({ providedIn: 'root' })
export class AgeStore extends ImmutableStore<TAge> {
  constructor() {
    super({
      cloneAndMutateFunc: produce as unknown as MutationFn<TAge>,
      enableLogging: true,
      initialState: new Age(),
      name: 'RegistrationStore',
      plugins: [useDevtools(), useStoreHistory(), useStorePersistence()],
    });
  }
}
```

**Comment by st-clair-clarke** [comment]
Great! Thank you. This works. Please incorporate in the docs.

When using structura produce(), there is no need to use `[immerable] = true` in Typescrpt classes? This only applies if produce is imported from immer?

**Comment by giusepperaso** [comment]
Hi everyone, sorry for being so late

Starting from version 0.11.4 the first example will compile so I think the issue is solved

Thanks @st-clair-clarke for finding that and @zuriscript for suggesting a workaround in the meanwhile :)

(The problem was due to a wrong type inference which made it very difficult to call methods on drafts and drafted objects)

I can confirm @st-clair-clarke that with structura you do not need [immerable] = true, this only applies if you use Immer; structura would just ignore that property and draft anyway

I will add some examples in the docs for produce in the next few days

**Comment by st-clair-clarke** [comment]
Cheers! Thank you. Love the lib!

**Comment by giusepperaso** [comment]
After that I noticed another related issue on Dates: if you tried to use methods like getTime or setDate inside a producer then it would trigger an error at runtime

This is now solved in version 0.11.6
